### ISSUE TITLE
change ephemeral progress output?

### ISSUE THREAD

**Issue by jpritikin** [OpenMx/OpenMx#290]
@khusmann In #77 it was suggested that we might change the output from "# evaluations" to "# of eval per time"

I'm opening a new issue to track this idea.

**Comment by khusmann** [comment]
As a user I only use # of evals to estimate the speed and check how long it's been running. So maybe something like this?

`[hh:mm:ss] MxComputeGradientDescent(SLSQP) 40eval/sec LL=318876 ΔLL=-1.955e+04`

**Comment by tbates** [comment]
This would be pleasant.
Nice to keep things in non-scientific where possible (maybe set a threshold of e+-6?), and marking thousands helps parsing:

```
[hh:mm:ss] MxComputeGradientDescent(SLSQP): 40eval/sec  LL = 318,876 Δ = -19,550
```

**Comment by khusmann** [comment]
Related: when I'm running models from non-interactive R processes, I can't seem to get progress outputs even when I run with silent=F (is there another option I'm missing?)

What would be really slick is if there was a way to connect to / send messages to silent openmx processes (thru process, pipe, socket, w/e) to get status updates, force checkpoints, etc. I find I'm often in a position where I've got a model that's been running on a node on the cluster and I have no way to tell if it something is wrong or it's making good progress other than to look at the last time a checkpoint was fsync'd (which who knows w nfs) and the # of cores currently being used. Yeah, it's quite the luxurious feature request but I figured I'd put it here while I was thinking about it to see if it might be useful to anyone else. If so, maybe I'll look into adding it myself one of these days :)

**Comment by jpritikin** [comment]
> Related: when I'm running models from non-interactive R processes, I can't seem to get progress outputs even when I run with silent=F (is there another option I'm missing?)

Yeah, OpenMx conditions on `interactive()`. Why do you want non-interactive progress output?

> What would be really slick is if there was a way to connect to / send messages to silent openmx processes ...

Yeah, this would be awesome. I agree.

**Comment by khusmann** [comment]
> Yeah, OpenMx conditions on `interactive()`. Why do you want non-interactive progress output?

I run big batches of large models on a cluster through scripts that call R in non-interactive mode. stderr/stdout are directed to files which I can read later. When something is going wrong in OpenMx, I have to manually get an interactive R process running on one of my nodes and run things there to get a sense of what's going on. It would make debugging/monitoring easier if I could sometimes turn on progress output in non-interactive R processes (and have OpenMx print status using \n instead of \r) so all I need to do is examine the stderr/stdout files rather than going to the trouble of getting an interactive R up and running.

**Comment by jpritikin** [comment]
> I run big batches of large models on a cluster

Wouldn't it be better to always enable checkpointing?

**Comment by khusmann** [comment]
> Wouldn't it be better to always enable checkpointing?

I would, if mxTryHard() supported checkpointing (#282). So maybe that's the larger issue here.

Still though, sometimes it is nice to have non-interactive progress output from OpenMx, like if I'm running a single job from my batch script locally and wanting to easily keep tabs on how it's going (rather having to always muck thru the checkpoint file and compare timestamps). Sure, not mission critical, but the option would be a nice ux boost in some use cases.

**Comment by mcneale** [comment]
I think mxTryHard should support checkpointing.  I’m surprised that it doesn’t, tbh, as it’s some of the best diagnostic information for difficult-to-optimize problems.  

For batch jobs, having checkpointing does, IMO, provide a window into the progress of a long-running batch job.  I like to use less to view the checkpoint file, as jumping to the end (G), and watching new lines of the file appear (F), plus being able to search backwards and forwards are quick and easy to use.  Reading it into R or Excel can provide further options (e.g., graphical) for viewing progress.

**Comment by khusmann** [comment]
>  I like to use less to view the checkpoint file

I do the same thing, but with a lot of free vars the it can be a long distance from the current LL value and the timestamp and so hard to quickly get a sense of how fast things are converging.  But sure, it's not too bad, and I suppose could just write a little awk script and run it with watch for easier monitoring. So yeah, the much larger issue is lack of checkpointing in mxTryHard.

**Comment by khusmann** [comment]
Here's a simple one-liner to make progress/checkpoint monitoring easier:

`watch 'cat CHECKPOINTFILE.omx | awk -F "\t" -vOFS=, '\''{print $1,$3,$4,$5,$NF}'\'' | tail | column -s , -t'`

output:

```
Every 2.0s: cat theory/theory.omx | awk -F "\t" -vOFS=, '{print $1,$3,$4,$5,$NF}' | tail | column -s , -t                                           Tue Mar 30 15:47:51 2021

OpenMxContext                    OpenMxEvals  iterations  timestamp                objective
MxComputeGradientDescent(SLSQP)  0            0           Mar 30 2021 03:30:26 PM  509512.4204
MxComputeGradientDescent(SLSQP)  2            1           Mar 30 2021 03:35:26 PM  509261.6446
MxComputeGradientDescent(SLSQP)  6905         2           Mar 30 2021 03:40:40 PM  509024.0151
MxComputeGradientDescent(SLSQP)  13810        3           Mar 30 2021 03:45:53 PM  507776.7203

```
